subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/NestedBeanResolver.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.properties;
+
+import com.google.common.collect.Iterables;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+public class NestedBeanResolver<T extends BeanNode<T>> {
+
+    public void resolve(T initial, NestedBeanContext<T> context) {
```

Comment:
"resolve" is a very overloaded term in Gradle's codebase. Can we fid a better name for this type and method? Something that reflects the nature of what's happening here, i.e. that we traverse to the bottom of an iterable "hierarchy". Maybe some Javadoc wouldn't hurt either.

Reply:
I went for `NestedPropertyUtil.collectNestedProperties`.